### PR TITLE
WIP: Refactor LEDControl into Controllers and Devices

### LINKED ISSUE #665: refactor LedControl
After trying to tweak `LedControl.js` to work with a [24-segment, bi-color bar-graph](https://www.adafruit.com/products/1721), I've come to the conclusion that the code needs some refactoring, because it scales poorly.
- if you are attempting to use the `HT16K33` controller in the context of a bar graph (since that's the chip the bar graph backpack uses), certain functions should not be available, and others (or rather, one, `setBar()`) should be.  
- it is difficult to reuse code in a controller given the common pattern of defining controllers; no "subcontrollers" are allowed
- Adding new functions requires explicitly adding them to the context of the `LedControl` instance (which _could_ be done via something like `_.mixin()` or `_.assign()`)
- controllers are not sufficiently encapsulated as to offer unique implementations for a display using a unique IC; the HT16K33 chip can drive all sorts of displays, not just traditional matrices or arrays of matrices.
- Matrix dimension restrictions should be shelved except in the case of hardware-specific controllers.  I don't need the code to tell me that driving six LEDs with a 74595 is not allowed!
- I don't understand what a `DEFAULT` controller buys us, if `LedControl.prototype` is non-empty.

To solve this, I would expect to use either inheritance or composition to extend the `LedControl` object with the proper functions.

Three (3) solutions follow.

Using inheritance:
- The `DEFAULT` controller's general-purpose methods are folded into `LedControl.prototype`, if any.  For all intents & purposes, `LedControl` is abstract.
- The `HT16K33` pseudoclass would be a static property of `LedControl`, and extend `LedControl`, providing specific implementation for this chip.   It contains no code which assumes any given matrix configuration.
- `Matrix`, `Digit` (or `SevenSegment`), `BarGraph` and even `Alphanumeric` would be static, class-level psuedoclasses in `HT16K33`, containing code specific to each implementation.  The `Adafruit_LEDBackpack` lib has implementation for all of these.
- The `ShiftRegister74595` (for lack of a better name) pseudoclass would do likewise.
- The organization scheme here is simply for sake of organization and is not required.
- Use the factory pattern to grab the class you want:
  
  ``` js
  LedControl.create = function create(type, controller, opts) {
    return new LedControl[controller][type](opts);
  };
  
  // alternatively:
   LedControl.createMatrix = function createMatrix(controller, opts) {
     return new LedControl[controller].Matrix(opts);
  };
  
  LedControl.createBarGraph = function createBarGraph(controller, opts) {
     return new LedControl[controller].BarGraph(opts);
  };
  // etc
  ```

Using composition:
- Fold the general-purpose functions from `DEFAULT` back into `LedControl.prototype`, if any.
- Create an data structure similar to the above, but use static objects instead of classes.
- Require a `type` string param upon instantiation, or default to `'matrix'`.  Require a `controller` param like we currently do, and default to `'74595'`.  Or use the factory pattern above.
- These two params are used to find the implementation, and within the `LedControl` constructor, set this value to `this.imp`. Since controllers and "types" would each have implementations, we probably have to mix them together.
- In `LedControl.prototype`, for each unique function in each implementation, expose a function which delegates to the same function in `this.imp`.
- If the function doesn't exist in the implementation, we can toss an error, or let Node throw a `TypeError`.

Using more mixins:
- Ginsu the mixins.  A controller-level mixin object would have methods common to each "type" for a given controller, and each type would have its own implementation(s).  For example:
  
  ``` js
  LedControl.HT16K33 = {
    commonFunc: function() { 
      // ...
    },
    matrix: {
      matrixyThing: function() { 
        // ...
      }
    },
    bargraph: {
      bargraphOnlyFunc: function() { 
        // ...
      }
    }
  };
  ```
- Change the mixin strategy.  When mixing in a controller, mix in _functions only_, then descend into the "type" object, and mix that in too.  Do this via a loop (or `_.mixin()`)  instead of explicitly defining the function properties on the context (see `LedControl` constructor).

Given that prototypal inheritance is fast, I don't see either of the first two strategies having a significant performance drawback.  Mixins ("parasitic inheritance"?) _are_ slower, though I don't know if it matters.

The "leaf" level implementations ("types" within controllers) should enforce the matrix dimensions.  Seven-segment displays will have a matrix of `n * 8` where `n` is the number of displays, and if we assume bar graph displays to have 12 bars, `n * 12`.  Perhaps this would necessitate an `init()` function for each "leaf" implementation, _if_ inheritance is not used.

I'd recommend using inheritance here, because the implementations are not guaranteed to have the same set of methods--it won't scale if a new method or controller is added.  Additionally, we avoid having to explicitly call an initialization function.  To me, composition makes more sense when you have a strict set of functions to be implemented, and this can be enforced (think interfaces in Java).

Anyway, since I'm in here trying to get my damn bar graph to work, I can do this and at least present you the result.

### PR DESCRIPTION
**Pull request by ericandrewlewis**
Work in progress to fix #665. Feedback welcome :smile:. 

I'm following patterns existing in the Motor module, separating logic out into "devices" (e.g. 8x8 bicolor) and "controllers" (e.g. `HT16K33`). 

@dtex @rwaldron @boneskull can one of you confirm this is a good path forward? Happy to continue!

### CODE REVIEW

**Comment by ericandrewlewis** [comment]
Going to continue with this ;) Will post here with updates, any feedback as progress developers appreciated!

**Comment by rwaldron** [comment]
I have not forgotten about this, just super busy. I will put together a test rig hopefully this weekend

File: lib/led/ledcontrol.js [hunk 1]
```diff
@@ -59,9 +59,8 @@ var priv = new Map(),
  *                                    Only for use with the HT16K33 controller.
  */
 function LedControl(opts) {
-
   Board.Component.call(
-    this, opts = Board.Options(opts)
+    this, this.opts = opts = Board.Options(opts)
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: Don't assign `opts` to `this.opts`. Some classes have this and it's a mistake that needs to be rectified

File: lib/led/ledcontrol.js [hunk 2]
```diff
@@ -80,18 +79,7 @@ function LedControl(opts) {
...
-  this.led = controller.led;
-  this.row = controller.row;
-  this.scanLimit = controller.scanLimit;
-  this.send = controller.send;
-  this.sendDigit = controller.sendDigit;
-  this.initialize = controller.initialize;
-
-  // controller specific op codes
-  this.OP = controller.OP;
+  Board.Controller.call(this, controller, opts);
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: I believe Controllers are setup after Device? cc @dtex

**Comment by ericandrewlewis** [comment]
@rwaldron no worries! I've been meaning to give a plain-text status update since this is such a huge refactor. Waiting for a component order to come in so I can test all the devices.

File: lib/led/ledcontrol.js (hunk 2 above)
Comment (ericandrewlewis) [inline lib/led/ledcontrol.js]: Yep, that's [the order the motor module takes](https://github.com/rwaldron/johnny-five/blob/master/lib/motor.js#L529-L544), I'll update accordingly.

Comment (dtex) [inline lib/led/ledcontrol.js]: Sorry, totally missed the mention. Yes, controller overrides device so :+1:

File: lib/led/ledcontrol.js (hunk 1 above)
Comment (dtex) [inline lib/led/ledcontrol.js]: reflectanceArray and motor are the offenders. I will fix the latter.

**Comment by ericandrewlewis** [comment]
Adjusted the commit based on feedback.

To recap what's going on before going further in this PR: what were previously called "controllers" (e.g. HT16K33) are being split into devices ("matrix display" and "seven-segment display") and controllers (HT16K33). Both of these will be mixins on an LedControl instance that utilizes `Object.defineProperties()`. 

Devices should track state of themselves. e.g. the matrix display tracks which pixels are on and off. 

The role of the controller is to write to the physical controller. 

This generally means that a matrix device will contain methods like `led()` (to turn on an led in a matrix) and `row()` (to set the state of a row in a matrix).

I notice that the 7219 is [written to one row at a time](https://github.com/rwaldron/johnny-five/blob/fd560084c3a150e5d08f50766d7fba10cf425e5c/lib/led/ledcontrol.js#L965-L991) and the HT16K33 is [written to all at once](https://github.com/rwaldron/johnny-five/blob/fd560084c3a150e5d08f50766d7fba10cf425e5c/lib/led/ledcontrol.js#L680-L688). To extract the `led()` method from both of these controllers into a single method on the "matrix" device will require some thought, since in the `led()` method we'll want to set state and then write that to the display. Should we create a new `writeDisplay()` method on the 7219 controller?

**Pull request closed without merging**